streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -623,6 +624,16 @@ private void closeGlobalStreamThread() {
     public synchronized boolean close(final long timeout, final TimeUnit timeUnit) {
         log.debug("{} Stopping Kafka Stream process.", logPrefix);
 
+        // do not close during state changes, e.g., there are corner cases when all
+        // threads die and the final thread to die calls onChange() is called to set the instances state.
+        // Within the instance callback, the user might call close() on the instance, which will attempt to join()
+        // the thread that started. So we have a circular loop (deadlock). We cannot stop the user from calling close()
+        // however we can prevent it from occurring.
+        if (inTheMiddleOfStateChange) {
```

Comment:
Thanks for bringing this up @dguy @enothereska , thinking about this more I feel it is Okay to state that users should not call any public APIs of the `KafkaStreams` object inside this callback, if they do then undefined behavior.